router/route/user.go

```diff
@@ -121,15 +122,104 @@ func retrieveUser(w http.ResponseWriter, r *http.Request) (int, error) {
 func retrieveUserScore(w http.ResponseWriter, r *http.Request) (code int, err error) {
 	username := mux.Vars(r)["username"]
 	score, err := model.GetUserScore(username)
+	err = json.NewEncoder(w).Encode(score)
 	if err != nil {
 		return http.StatusInternalServerError, causerr.New(err, "")
 	}
+	return http.StatusOK, nil
+}
 
-	err = json.NewEncoder(w).Encode(score)
+// putFriend returns
+func putFriend(w http.ResponseWriter, r *http.Request) (int, error) {
+	s, err := session.Get(r, "roaster_auth")
 	if err != nil {
 		return http.StatusInternalServerError, causerr.New(err, "")
 	}
 
+	username, ok := s.Values["username"].(string)
+	if !ok || username == "" {
+		return http.StatusNoContent, nil
+	}
+
+	f := model.Friend{}
+
+	err = json.NewDecoder(r.Body).Decode(&f)
+	if err != nil {
+		log.Println(err)
+		return http.StatusBadRequest, causerr.New(err,
+			"unable to decode request")
+	}
+	if f.Friend == "" {
+		return http.StatusBadRequest, causerr.New(
+			errors.New("friend is empty"),
+			"Missing friend parameter in URI")
+	}
+
+	err = model.PutFriend(username, f.Friend)
+	if err != nil {
+		log.Println(err)
+		if pgerr, ok := err.(*pq.Error); ok {
+			log.Println(pgerr.Constraint)
+			if pgerr.Constraint == "friend_relation_uq" {
+				return http.StatusConflict, causerr.New(err, "User already has this friend")
+			}
+			if pgerr.Constraint == "username_fk" {
+				return http.StatusBadRequest, causerr.New(err, "No user registered, are you logged in?")
+			}
+		}
+		return http.StatusInternalServerError, causerr.New(err, "")
+	}
+	return http.StatusOK, nil
+}
+
+// retrieveFriends returns
+func retrieveFriends(w http.ResponseWriter, r *http.Request) (int, error) {
+	vars := mux.Vars(r)
+	username := vars["username"]
+
+	if username == "" {
+		return http.StatusBadRequest, causerr.New(
+			errors.New("missing username for lookup"),
+			"Missing username for lookup")
+	}
+	friends, err := model.GetFriends(username)
+	if err != nil {
+		return http.StatusInternalServerError, causerr.New(err, "")
```

Comment:
Hmm, this could also the a `404` right? But w/e we're stressed.

Reply:
Le Optionale

Reply:
Look 3 rows down, there I handle the case and send the appropriate "204 No content". Since the request is successful, no internal server errors but the user has 0 friends, hence no content. Seems OK to you?

Reply:
You're right, the DB query won't return any error if the user doesn't exist.